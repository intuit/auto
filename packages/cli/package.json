{
  "name": "auto",
  "bin": "dist/bin/auto.js",
  "main": "dist/bin/index.js",
  "types": "dist/index.d.ts",
  "description": "CLI tools to help facilitate semantic versioning based on GitHub PR labels",
  "version": "10.46.0",
  "license": "MIT",
  "author": {
    "name": "Andrew Lisowski",
    "email": "lisowski54@gmail.com"
  },
  "pkg": {
    "scripts": "../../plugins/**/dist/**/*.js"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/intuit/auto"
  },
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=10.x"
  },
  "keywords": [
    "automation",
    "semantic",
    "release",
    "github",
    "labels",
    "automated",
    "continuos integration",
    "changelog",
    "cli"
  ],
  "scripts": {
    "build": "tsc -b",
    "start": "npm run build -- -w",
    "lint": "eslint src --ext .ts",
    "test": "jest --maxWorkers=2",
    "bundle": "yarn inject-version && yarn package && yarn gzip",
    "inject-version": "node scripts/inject-version.js",
    "package": "rimraf binary && pkg . --out-path binary",
    "gzip": "ls binary/auto* | xargs gzip"
  },
  "dependencies": {
    "@auto-it/core": "file:../core",
    "@auto-it/npm": "file:../../plugins/npm",
    "@auto-it/released": "file:../../plugins/released",
    "@auto-it/version-file": "file:../../plugins/version-file",
    "await-to-js": "^3.0.0",
    "chalk": "^4.0.0",
    "command-line-application": "^0.10.1",
    "endent": "^2.1.0",
    "module-alias": "^2.2.2",
    "signale": "^1.4.0",
    "terminal-link": "^2.1.1",
    "tslib": "2.1.0"
  },
  "devDependencies": {
    "@types/signale": "^1.2.1",
    "pkg": "^5.8.0"
  }
}
