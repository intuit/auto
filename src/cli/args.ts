import * as fs from 'fs';
import * as path from 'path';

import chalk from 'chalk';
import commandLineArgs from 'command-line-args';
import commandLineUsage from 'command-line-usage';
import signale from 'signale';

const p = chalk.hex('#870048');
const y = chalk.hex('#F1A60E');
const r = chalk.hex('#C5000B');
const g = chalk.hex('#888888');

// prettier-ignore
const logo = `
      ${y('_________')}
     ${p('/')}${y('\\       /')}${r('\\')}                   _______ _     _ _______  _____
    ${p('/')}  ${y('\\_____/')}  ${r('\\')}                  |_____| |     |    |    |     |
   ${p('/   /')}     ${r('\\   \\')}                 |     | |_____|    |    |_____|
  ${p('/___/')} \\▔▔\\ ${r(' \\___\\')}
  ${g('\\   \\')}  \\_/  ${g('/   /')}     ______ _______        _______ _______ _______ _______
   ${g('\\   \\')}     ${g('/   /')}     |_____/ |______ |      |______ |_____| |______ |______
    ${g('\\   ▔▔▔▔▔   /')}      |    \\_ |______ |_____ |______ |     | ______| |______
     ${g('\\         /')}
      ${g('▔▔▔▔▔▔▔▔▔ ')}
`;

const help: commandLineUsage.OptionDefinition = {
  name: 'help',
  alias: 'h',
  type: Boolean
};

const version = {
  name: 'version',
  alias: 'V',
  type: Boolean,
  description: "Display auto-release-cli's version"
};

const mainDefinitions = [
  { name: 'command', type: String, defaultOption: true },
  {
    ...help,
    description: 'Display the help output. Works on each command as well'
  },
  version
];

const defaultOptions = [
  {
    ...help,
    description: 'Display the help output for the command',
    group: 'misc'
  },
  {
    name: 'verbose',
    alias: 'v',
    type: Boolean,
    description: 'Show some more logs',
    group: 'misc'
  },
  {
    name: 'very-verbose',
    alias: 'w',
    type: Boolean,
    description: 'Show a lot more logs',
    group: 'misc'
  },
  {
    name: 'repo',
    type: String,
    description:
      'The repo to set the status on. Defaults to looking in the package definition for the platform',
    group: 'misc'
  },
  {
    name: 'owner',
    type: String,
    description:
      'The owner of the GitHub repo. Defaults to reading from the package definition for the platform',
    group: 'misc'
  },
  {
    name: 'github-api',
    type: String,
    description: 'GitHub API to use',
    group: 'misc'
  },
  {
    name: 'plugins',
    type: String,
    multiple: true,
    description: 'Plugins to load auto-release with. (defaults to just npm)',
    group: 'misc'
  }
];

const pr: commandLineUsage.OptionDefinition = {
  name: 'pr',
  type: Number,
  description: 'The pull request number you want the labels of',
  group: 'main'
};

const dryRun: commandLineUsage.OptionDefinition = {
  name: 'dry-run',
  alias: 'd',
  type: Boolean,
  description: 'Report what command will do but do not actually do anything',
  group: 'main'
};

const url: commandLineUsage.OptionDefinition = {
  name: 'url',
  type: String,
  description: 'URL to associate with this status',
  group: 'main'
};

const noVersionPrefix: commandLineUsage.OptionDefinition = {
  name: 'no-version-prefix',
  type: Boolean,
  description: "Use the version as the tag without the 'v' prefix",
  group: 'main'
};

const jira: commandLineUsage.OptionDefinition = {
  name: 'jira',
  type: String,
  description: 'Jira base URL',
  group: 'main'
};

const name: commandLineUsage.OptionDefinition = {
  name: 'name',
  type: String,
  description:
    'Git name to commit and release with. Defaults to package definition for the platform',
  group: 'main'
};

const email: commandLineUsage.OptionDefinition = {
  name: 'email',
  type: String,
  description:
    'Git email to commit with. Defaults to package definition for the platform',
  group: 'main'
};

const context: commandLineUsage.OptionDefinition = {
  name: 'context',
  type: String,
  description: 'A string label to differentiate this status from others',
  group: 'main'
};

const message: commandLineUsage.OptionDefinition = {
  name: 'message',
  group: 'main',
  type: String,
  alias: 'm'
};

const skipReleaseLabels: commandLineUsage.OptionDefinition = {
  name: 'skip-release-labels',
  type: String,
  group: 'main',
  multiple: true,
  description:
    "Labels that will not create a release. Defaults to just 'skip-release'"
};

interface ICommand {
  name: string;
  summary: string;
  options?: commandLineUsage.OptionDefinition[];
  require?: Flags[];
  examples: any[];
}

const commands: ICommand[] = [
  {
    name: 'init',
    summary: 'Interactive setup for most configurable options',
    examples: ['{green $} auto init'],
    options: [
      {
        name: 'only-labels',
        type: Boolean,
        group: 'main',
        description:
          'Only run init for the labels. As most other options are for advanced users'
      },
      dryRun
    ]
  },
  {
    name: 'create-labels',
    summary: "Create your project's labels on github.",
    examples: ['{green $} auto create-labels'],
    options: [...defaultOptions, dryRun]
  },
  {
    name: 'label',
    summary: 'Get the labels for a pull request',
    options: [
      { ...pr, description: `${pr.description} (defaults to last merged PR)` },
      ...defaultOptions
    ],
    examples: ['{green $} auto label --pr 123']
  },
  {
    name: 'pr-check',
    summary: 'Check that a pull request has a SemVer label',
    require: ['pr', 'url'],
    options: [
      pr,
      url,
      dryRun,
      {
        ...context,
        defaultValue: 'ci/pr-check'
      },
      skipReleaseLabels,
      ...defaultOptions
    ],
    examples: [
      '{green $} auto pr-check --pr 32 --url http://your-ci.com/build/123'
    ]
  },
  {
    name: 'pr',
    summary: 'Set the status on a PR commit',
    require: ['state', 'url', 'description', 'context'],
    options: [
      {
        name: 'sha',
        type: String,
        group: 'main',
        description:
          'Specify a custom git sha. Defaults to the HEAD for a git repo in the current repository'
      },
      { ...pr, description: 'PR to set the status on' },
      url,
      {
        name: 'state',
        type: String,
        group: 'main',
        description:
          "State of the PR. ['pending', 'success', 'error', 'failure']"
      },
      {
        name: 'description',
        type: String,
        group: 'main',
        description: 'A description of the status'
      },
      {
        name: 'context',
        type: String,
        group: 'main',
        description: 'A string label to differentiate this status from others'
      },
      dryRun,
      ...defaultOptions
    ],
    examples: [
      `{green $} auto pr \\\\ \n   --pr 32 \\\\ \n   --state pending \\\\ \n   --description "Build still running..." \\\\ \n   --context build-check`
    ]
  },
  {
    name: 'version',
    summary: 'Get the semantic version bump for the given changes.',
    options: [
      {
        name: 'only-publish-with-release-label',
        type: Boolean,
        description: "Only bump version if 'release' label is on pull request",
        group: 'main'
      },
      skipReleaseLabels,
      ...defaultOptions
    ],
    examples: [
      {
        desc: 'Get the new version using the last release to head',
        example: '{green $} auto version'
      },
      {
        desc: 'Skip releases with multiple labels',
        example: '{green $} auto version --skip-release-labels documentation CI'
      }
    ]
  },
  {
    name: 'changelog',
    summary: "Prepend release notes to 'CHANGELOG.md'",
    options: [
      dryRun,
      noVersionPrefix,
      name,
      email,
      jira,
      {
        name: 'from',
        type: String,
        group: 'main',
        description:
          'Tag to start changelog generation on. Defaults to latest tag.'
      },
      {
        name: 'to',
        type: String,
        group: 'main',
        description: 'Tag to end changelog generation on. Defaults to HEAD.'
      },
      {
        ...message,
        description:
          "Message to commit the changelog with. Defaults to 'Update CHANGELOG.md [skip ci]'"
      },
      ...defaultOptions
    ],
    examples: [
      {
        desc: 'Generate a changelog from the last release to head',
        example: '{green $} auto changelog'
      },
      {
        desc: 'Generate a changelog across specific versions',
        example: '{green $} auto changelog --from v0.20.1 --to v0.21.0'
      }
    ]
  },
  {
    name: 'release',
    summary: 'Auto-generate a github release',
    options: [
      dryRun,
      noVersionPrefix,
      name,
      email,
      jira,
      {
        name: 'use-version',
        type: String,
        group: 'main',
        description:
          'Version number to publish as. Defaults to reading from the package definition for the platform.'
      },
      {
        name: 'slack',
        alias: 's',
        type: String,
        group: 'main',
        description:
          'Url to post a slack message to about the release. Make sure the SLACK_TOKEN environment variable is set.'
      },
      ...defaultOptions
    ],
    examples: ['{green $} auto release']
  },
  {
    name: 'comment',
    summary: 'Comment on a pull request with a markdown message',
    require: ['pr', 'message'],
    options: [
      pr,
      context,
      { ...message, description: 'Message to post to comment' },
      dryRun,
      ...defaultOptions
    ],
    examples: [
      '{green $} auto comment --pr 123 --comment "# Why you\'re wrong..."'
    ]
  },
  {
    name: 'shipit',
    summary: 'Run the full auto-release project. Detects if in a lerna project',
    examples: ['{green $} auto shipit'],
    options: [...defaultOptions, dryRun]
  }
];

function filterCommands(allCommands: ICommand[], include: string[]) {
  return allCommands
    .filter(command => include.includes(command.name))
    .map(command => ({
      name: command.name,
      summary: command.summary
    }));
}

function printRootHelp() {
  const options = [
    { ...version, group: 'misc' },
    ...mainDefinitions,
    ...defaultOptions
  ];
  options.forEach(option => styleTypes({} as ICommand, option));

  const usage = commandLineUsage([
    {
      content: logo.replace(/\\/g, '\\\\'),
      raw: true
    },
    {
      content:
        'Generate releases based on semantic version labels on pull requests'
    },
    {
      header: 'Synopsis',
      content: '$ auto <command> <options>'
    },
    {
      header: 'Setup Commands',
      content: filterCommands(commands, ['init', 'create-labels'])
    },
    {
      header: 'Release Commands',
      content: filterCommands(commands, [
        'release',
        'version',
        'changelog',
        'shipit'
      ])
    },
    {
      header: 'Pull Request Interaction Commands',
      content: filterCommands(commands, ['label', 'pr-check', 'pr', 'comment'])
    },
    {
      header: 'Global Options',
      optionList: options,
      group: 'misc'
    }
  ]);

  console.log(usage);
}

function printCommandHelp(command: ICommand) {
  const sections: commandLineUsage.Section[] = [
    {
      header: `auto ${command.name}`,
      content: command.summary
    }
  ];

  if (command.options) {
    const hasLocalOptions = command.options.filter(
      option => option.group === 'main'
    );

    if (hasLocalOptions.length > 0) {
      sections.push({
        header: 'Options',
        optionList: command.options,
        group: 'main'
      });
    }

    const hasGlobalOptions = command.options.filter(
      option => option.group === 'misc'
    );

    if (hasGlobalOptions.length > 0) {
      sections.push({
        header: 'Global Options',
        optionList: command.options,
        group: 'misc'
      });
    }
  }

  sections.push({
    header: 'Examples',
    content: command.examples,
    raw: command.name === 'pr'
  });

  console.log(commandLineUsage(sections));
}

function printVersion() {
  const packagePath = path.join(__dirname, '../../package.json');
  const packageJson = JSON.parse(fs.readFileSync(packagePath, 'utf8'));
  console.log(`v${packageJson.version}`);
}

function styleTypes(
  command: ICommand,
  option: commandLineUsage.OptionDefinition
) {
  const isRequired =
    command.require && command.require.includes(option.name as keyof ArgsType);

  if (isRequired && option.type === Number) {
    option.typeLabel =
      '{rgb(173, 216, 230) {underline number}} [{rgb(254,91,92) required}]';
  } else if (option.type === Number) {
    option.typeLabel = '{rgb(173, 216, 230) {underline number}}';
  }

  if (isRequired && option.type === String) {
    option.typeLabel =
      '{rgb(173, 216, 230) {underline string}} [{rgb(254,91,92) required}]';
  } else if (option.multiple && option.type === String) {
    option.typeLabel = '{rgb(173, 216, 230) {underline string[]}}';
  } else if (option.type === String) {
    option.typeLabel = '{rgb(173, 216, 230) {underline string}}';
  }
}

export default function parseArgs(testArgs?: string[]) {
  const mainOptions = commandLineArgs(mainDefinitions, {
    stopAtFirstUnknown: true,
    camelCase: true,
    argv: testArgs
  });
  const argv = mainOptions._unknown || [];
  const command = commands.find(c => c.name === mainOptions.command);

  if (!command && mainOptions.version) {
    return printVersion();
  }

  if (!command) {
    return printRootHelp();
  }

  const options = command.options || [];

  options.map(option => styleTypes(command, option));

  if (mainOptions.help) {
    return printCommandHelp(command);
  }

  const autoOptions: ArgsType = {
    command: mainOptions.command,
    ...commandLineArgs(options, { argv, camelCase: true })._all
  };

  if (command.require) {
    const missing = command.require
      .filter(
        option =>
          !autoOptions.hasOwnProperty(option) ||
          // tslint:disable-next-line strict-type-predicates
          autoOptions[option as keyof ArgsType] === null
      )
      .map(option => `--${option}`);
    const multiple = missing.length > 1;

    if (missing.length > 0) {
      printCommandHelp(command);
      signale.error(
        `Missing required flag${multiple ? 's' : ''}: ${missing.join(', ')}`
      );
      return process.exit(1);
    }
  }

  return autoOptions;
}

interface IAuthorArgs {
  name?: string;
  email?: string;
}

interface IRepoArgs {
  owner?: string;
  repo?: string;
}
interface ILogArgs {
  verbose?: boolean;
  veryVerbose?: boolean;
}

export interface IInitCommandOptions {
  onlyLabels?: boolean;
  dryRun?: boolean;
}

export interface ICreateLabelsCommandOptions {
  dryRun?: boolean;
}

export interface ILabelCommandOptions {
  pr?: number;
}

export interface IPRCheckCommandOptions {
  pr: number;
  url?: string;
  skipReleaseLabels?: string[];
  context?: string;
  dryRun?: boolean;
}

export interface IPRCommandOptions {
  sha?: string;
  pr?: number;
  url: string;
  state: 'pending' | 'success' | 'error' | 'failure';
  description: string;
  context: string;
  dryRun?: boolean;
}

export interface IVersionCommandOptions {
  skipReleaseLabels?: string[];
  onlyPublishWithReleaseLabel?: boolean;
}

export interface IChangelogOptions extends IAuthorArgs {
  noVersionPrefix?: boolean;
  jira?: string;
  dryRun?: boolean;
  from?: string;
  to?: string;
  message?: string;
}

export interface IReleaseOptions extends IAuthorArgs {
  noVersionPrefix?: boolean;
  jira?: string;
  dryRun?: boolean;
  useVersion?: string;
  slack?: string;
}

export interface ICommentCommandOptions {
  pr: number;
  message: string;
  context?: string;
  dryRun?: boolean;
}

export interface IShipItCommandOptions {
  dryRun?: boolean;
}

type GlobalFlags = {
  command: string;
  githubApi?: string;
  plugins?: string[];
} & IRepoArgs &
  ILogArgs;

export type ArgsType = GlobalFlags &
  (
    | IInitCommandOptions
    | ICreateLabelsCommandOptions
    | ILabelCommandOptions
    | IPRCheckCommandOptions
    | IPRCommandOptions
    | ICommentCommandOptions
    | IReleaseOptions
    | IVersionCommandOptions
    | IShipItCommandOptions);

type Flags =
  | keyof GlobalFlags
  | keyof IInitCommandOptions
  | keyof ICreateLabelsCommandOptions
  | keyof ILabelCommandOptions
  | keyof IPRCheckCommandOptions
  | keyof IPRCommandOptions
  | keyof ICommentCommandOptions
  | keyof IReleaseOptions
  | keyof IVersionCommandOptions
  | keyof IShipItCommandOptions;
