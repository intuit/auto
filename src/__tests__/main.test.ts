import { IPRCommandOptions } from '../cli/args';
import main, { AutoRelease, run } from '../main';

test('throws error for unknown args', async () => {
  expect.assertions(1);

  try {
    await run({
      command: 'foo'
    });
  } catch (error) {
    expect(error).toEqual(new Error("idk what i'm doing."));
  }
});

test('throws exits for caught error', async () => {
  console.log = jest.fn() as any;
  process.exit = jest.fn() as any;

  await main({
    command: 'foo'
  });

  expect(process.exit).toHaveBeenCalledWith(1);
});

const defaults = {
  owner: 'foo',
  repo: 'bar',
  token: 'XXXX'
};

const labels = {
  major: 'Version: Major',
  patch: 'Version: Patch',
  minor: 'Version: Minor'
};

const search = jest.fn();
jest.mock('cosmiconfig', () => () => ({
  search
}));

describe('AutoRelease', () => {
  test('should use args', async () => {
    const auto = new AutoRelease({ command: 'init', ...defaults });
    await auto.loadConfig();
    expect(auto.githubRelease).toBeDefined();
  });

  test('should load config', async () => {
    search.mockReturnValueOnce({ config: defaults });
    const auto = new AutoRelease({ command: 'init' });
    await auto.loadConfig();
    expect(auto.githubRelease).toBeDefined();
  });

  test.only('should fail if config contains extra property', async () => {
    search.mockReturnValueOnce({ config: { invalid: true } });
    const auto = new AutoRelease({ command: 'init' });
    return expect(auto.loadConfig()).rejects.toThrowError();
  });

  test('should use labels from config config', async () => {
    search.mockReturnValueOnce({
      config: { ...defaults, labels }
    });
    const auto = new AutoRelease({ command: 'init' });
    await auto.loadConfig();

    expect([...auto.semVerLabels!.values()]).toEqual([
      'Version: Major',
      'Version: Minor',
      'Version: Patch',
      'skip-release',
      'release',
      'prerelease'
    ]);
  });

  test('should add extra skip label', async () => {
    search.mockReturnValueOnce({
      config: {
        ...defaults,
        labels: {
          'skip-release': 'NOPE'
        }
      }
    });
    const auto = new AutoRelease({ command: 'init' });
    await auto.loadConfig();

    expect(auto.githubRelease!.releaseOptions.skipReleaseLabels).toEqual([
      'NOPE'
    ]);
  });

  describe('createLabels', () => {
    test('should throw when not initialized', async () => {
      search.mockReturnValueOnce({
        config: { ...defaults, labels }
      });
      const auto = new AutoRelease({ command: 'create-labels' });
      expect(auto.createLabels()).rejects.toBeTruthy();
    });

    test('should create the labels', async () => {
      search.mockReturnValueOnce({
        config: { ...defaults, labels }
      });
      const auto = new AutoRelease({ command: 'create-labels' });
      await auto.loadConfig();

      auto.githubRelease!.addLabelsToProject = jest.fn();
      await auto.createLabels();
      expect(auto.githubRelease!.addLabelsToProject).toMatchSnapshot();
    });
  });

  describe('label', () => {
    test('should throw when not initialized', async () => {
      search.mockReturnValueOnce({
        config: { ...defaults, labels }
      });
      const auto = new AutoRelease({ command: 'labels' });
      expect(auto.label({ pr: 13 })).rejects.toBeTruthy();
    });

    test('should get labels', async () => {
      const auto = new AutoRelease({ command: 'labels', ...defaults });
      await auto.loadConfig();

      const getLabels = jest.fn();
      auto.githubRelease!.getLabels = getLabels;
      getLabels.mockReturnValueOnce(['foo']);
      console.log = jest.fn();

      await auto.label({ pr: 13 });
      expect(console.log).toHaveBeenCalledWith('foo');
    });

    test('should get labels for last merged PR', async () => {
      const auto = new AutoRelease({ command: 'labels', ...defaults });
      await auto.loadConfig();

      const getPullRequests = jest.fn();
      auto.githubRelease!.getPullRequests = getPullRequests;
      getPullRequests.mockReturnValueOnce([
        { merged_at: true, labels: [{ name: 'foo' }, { name: 'bar' }] }
      ]);
      console.log = jest.fn();

      await auto.label();
      expect(console.log).toHaveBeenCalledWith('foo\nbar');
    });

    test('should do nothing when no last merge found', async () => {
      const auto = new AutoRelease({ command: 'labels', ...defaults });
      await auto.loadConfig();

      const getPullRequests = jest.fn();
      auto.githubRelease!.getPullRequests = getPullRequests;
      getPullRequests.mockReturnValueOnce([]);
      console.log = jest.fn();

      await auto.label();
      expect(console.log).not.toHaveBeenCalled();
    });
  });

  describe('pr', () => {
    let createStatus: jest.Mock;

    beforeEach(() => {
      createStatus = jest.fn();
    });

    const required: IPRCommandOptions = {
      url: 'https://google.com',
      state: 'pending',
      description: 'foo',
      context: 'bar'
    };

    test('should throw when not initialized', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      expect(auto.pr(required)).rejects.toBeTruthy();
    });

    test('should do nothing with dryRun', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();

      await auto.pr({ ...required, sha: '1234', dryRun: true });
      expect(createStatus).not.toHaveBeenCalled();
    });

    test('should use provided SHA', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      await auto.pr({ ...required, sha: '1234' });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          sha: '1234'
        })
      );
    });

    test('should use HEAD SHA', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      const getSha = jest.fn();
      auto.githubRelease!.getSha = getSha;
      getSha.mockReturnValueOnce('abc');

      await auto.pr({ ...required });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          sha: 'abc'
        })
      );
    });

    test('should use lookup SHA for PR', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      const getPullRequest = jest.fn();
      auto.githubRelease!.getPullRequest = getPullRequest;
      getPullRequest.mockReturnValueOnce({ data: { head: { sha: 'deep' } } });

      await auto.pr({ ...required, pr: 14 });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          sha: 'deep'
        })
      );
    });
  });

  describe('pr-check', () => {
    jest.setTimeout(10 * 1000);
    let createStatus: jest.Mock;

    beforeEach(() => {
      createStatus = jest.fn();
    });

    const required = {
      url: 'https://google.com'
    };

    test('should throw when not initialized', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      expect(auto.prCheck({ pr: 13, ...required })).rejects.toBeTruthy();
    });

    test('should do nothing with dryRun', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();

      await auto.prCheck({ ...required, pr: 13, dryRun: true });
      expect(createStatus).not.toHaveBeenCalled();
    });

    test('should catch errors', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      await auto.prCheck({ ...required, pr: 13 });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          state: 'error'
        })
      );
    });

    test('should error with no label', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      const getPullRequest = jest.fn();
      auto.githubRelease!.getPullRequest = getPullRequest;
      getPullRequest.mockReturnValueOnce({ data: { head: { sha: 'sha' } } });

      const getLabels = jest.fn();
      auto.githubRelease!.getLabels = getLabels;
      getLabels.mockReturnValueOnce([]);

      await auto.prCheck({ ...required, pr: 13 });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          description: 'No semver label!'
        })
      );
    });

    test('should pass with semver label', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      const getPullRequest = jest.fn();
      auto.githubRelease!.getPullRequest = getPullRequest;
      getPullRequest.mockReturnValueOnce({ data: { head: { sha: 'sha' } } });

      const getLabels = jest.fn();
      auto.githubRelease!.getLabels = getLabels;
      getLabels.mockReturnValueOnce(['major']);

      await auto.prCheck({ ...required, pr: 13 });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          description: 'CI - major'
        })
      );
    });

    test('should pass with skip release label', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      const getPullRequest = jest.fn();
      auto.githubRelease!.getPullRequest = getPullRequest;
      getPullRequest.mockReturnValueOnce({ data: { head: { sha: 'sha' } } });

      const getLabels = jest.fn();
      auto.githubRelease!.getLabels = getLabels;
      getLabels.mockReturnValueOnce(['major', 'skip-release']);

      await auto.prCheck({ ...required, pr: 13 });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          description: 'PR will not create a release'
        })
      );
    });

    test('should pass with skip release label', async () => {
      const auto = new AutoRelease({ command: 'pr', ...defaults });
      await auto.loadConfig();
      auto.githubRelease!.createStatus = createStatus;

      const getPullRequest = jest.fn();
      auto.githubRelease!.getPullRequest = getPullRequest;
      getPullRequest.mockReturnValueOnce({ data: { head: { sha: 'sha' } } });

      const getLabels = jest.fn();
      auto.githubRelease!.getLabels = getLabels;
      getLabels.mockReturnValueOnce(['major', 'release']);

      await auto.prCheck({ ...required, pr: 13 });
      expect(createStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          description: 'PR will create release once merged - major'
        })
      );
    });
  });

  describe('comment', () => {
    test('should throw when not initialized', async () => {
      const auto = new AutoRelease({ command: 'comment', ...defaults });
      expect(auto.comment({ pr: 10, message: 'foo' })).rejects.toBeTruthy();
    });

    test('should make a comment', async () => {
      const auto = new AutoRelease({ command: 'comment', ...defaults });
      await auto.loadConfig();

      const createComment = jest.fn();
      auto.githubRelease!.createComment = createComment;

      await auto.comment({ pr: 10, message: 'foo' });
      expect(createComment).toHaveBeenCalled();
    });
  });

  describe('version', () => {
    test('should throw when not initialized', async () => {
      const auto = new AutoRelease({ command: 'comment', ...defaults });
      expect(auto.version()).rejects.toBeTruthy();
    });

    test('should make a comment', async () => {
      const auto = new AutoRelease({ command: 'comment', ...defaults });
      await auto.loadConfig();

      const getSemverBump = jest.fn();
      auto.githubRelease!.getLatestRelease = jest.fn();
      auto.githubRelease!.getSemverBump = getSemverBump;
      getSemverBump.mockReturnValueOnce('patch');
      console.log = jest.fn();

      await auto.version();
      expect(console.log).toHaveBeenCalledWith('patch');
    });
  });

  describe('changelog', () => {
    test('should throw when not initialized', async () => {
      const auto = new AutoRelease({ command: 'comment', ...defaults });
      expect(auto.changelog()).rejects.toBeTruthy();
    });

    test('should do nothing on a dryRun', async () => {
      const auto = new AutoRelease({ command: 'comment', ...defaults });
      await auto.loadConfig();

      const addToChangelog = jest.fn();
      auto.githubRelease!.addToChangelog = addToChangelog;
      auto.githubRelease!.generateReleaseNotes = jest.fn();

      await auto.changelog({ from: 'v1.0.0', dryRun: true });
      expect(addToChangelog).not.toHaveBeenCalled();
    });

    test('should add to changelog', async () => {
      const auto = new AutoRelease({
        command: 'changelog',
        plugins: [],
        ...defaults
      });
      auto.hooks.getRepository.tap('test', () => ({ token: '1234' }));
      await auto.loadConfig();

      const addToChangelog = jest.fn();
      auto.githubRelease!.addToChangelog = addToChangelog;
      auto.githubRelease!.generateReleaseNotes = jest.fn();

      await auto.changelog({ from: 'v1.0.0' });
      expect(addToChangelog).toHaveBeenCalled();
    });
  });
});
