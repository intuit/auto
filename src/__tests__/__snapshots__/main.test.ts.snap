// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoRelease createLabels should create the labels 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Map {
        "major" => "major",
        "minor" => "minor",
        "patch" => "patch",
        "skip-release" => "skip-release",
        "release" => "release",
        "prerelease" => "prerelease",
        "internal" => "internal",
        "documentation" => "documentation",
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
